Author: Pietro Ariston Zubiri

Last Updated: 4/26/2025

## Project Background and Rationale

This data was scraped responsibly with Selenium from the official Glock website (https://eu.glock.com) on April 24, 2025. Only public data was accessed.
I chose to explore and analyse the active product range of Glock pistols for this exploaratory data analysis (EDA) project as I thought it would be interesting to apply data science methods to a domain I knew little about. Exploring unfamiliar industries is always a good way to sharpen up ones ability to ask questions and stay curious.  Special thanks to my friend Brandon for teaching me the basics of firearms to help kickstart this project.
I hope you find my work interesting.

## The Story of Glock

<p align="center">
  <img src="images/glock_art.jpg" alt="Gaston Glock and the G17 Prototype" width="800"/>
</p>

Gaston Glock was an Austrian engineer and businessman who founded Glock in 1963. Initially, the company specialized in manufacturing curtain rods and military equipment like knives and grenade casings. When Glock turned to firearms development in the early 1980s, he had no prior pistolsmithing experience — but his background in industrial fabrication and materials engineering became a key advantage. Instead of relying on traditional, handcrafted methods, Glock emphasized the use of modern machinery, automation, and polymer technology. This approach reduced production costs, minimized human error, and created consistent, high-quality parts with tight tolerances. His first firearm, the Glock 17, was a lightweight, durable, and highly reliable semi-automatic pistol. It quickly won the Austrian Army’s contract in 1982 and soon became a global phenomenon. Today, Glock pistols are the most widely used handpistols by law enforcement, military, and civilians around the world, valued for their rugged simplicity and unmatched reliability.

## Reputation of Glock

<br><p align="center">
  <img src="./images/glock22_gen5.jpg" alt="Glock-22 Gen 5" width="800"/>
</p>
Glock pistols have built a reputation for being rugged, reliable, and highly functional. When pistol owners think of a Glock, they often associate it with simplicity, toughness and practicality- a "no-frills" tool built to perform in any scenario. However, For many people unfamiliar with firearms, Glocks have gained a reputation as small, lightweight "plastic pistols" with limited power and range. This perception stems largely from popular media, where Glocks are often portrayed as token weapons or standard-issue pistols that lack the stopping power of larger firearms. Movies and video games like Counter-Strike have reinforced this image by positioning Glocks as token or starter weapons. In this project, we will explore and analyze the range of Glock firearms and debunk facts from fiction through statistical methods and learn a thing or two about Glock pistols.

## Initial Exploration and Data Quality Assessment
If you aren't here for the technical bits, skip over this section.

In [24]:
# Initializing dependencies

import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

current_dir = os.getcwd()
csv_path = os.path.join(current_dir, 'glock_dataset.csv')

# Define Glock-inspired colors
glock_orange = "#EF9730"
glock_black = "#000000"

# Create custom Glock style palette
glock_palette = {
    "axes.facecolor": glock_black,
    "figure.facecolor": glock_black,
    "text.color": "white",
    "axes.labelcolor": "white",
    "xtick.color": "white",
    "ytick.color": "white",
    "grid.color": "#333333",
    "axes.edgecolor": glock_orange,
    "axes.spines.right": False,
    "axes.spines.top": False
}

# Set the style
sns.set_theme(style="darkgrid")
sns.set_style("darkgrid", rc=glock_palette)

# Set color palette with Glock orange as primary color
sns.set_palette([glock_orange, "white", "#FFBC75", "#555555"])

# Set plot context for better sizing
sns.set_context("notebook", font_scale=1.2)

# Optional: Set matplotlib defaults to complement the style
plt.rcParams['axes.grid'] = True
plt.rcParams['grid.linestyle'] = ':'
plt.rcParams['grid.linewidth'] = 0.5

In [2]:
# Importing data

data_raw = pd.read_csv(csv_path)
data = data_raw.copy()

data.head(3)

,name,description,url,Caliber,System,Mag. Capacity,Barrel Length,Trigger Pull***,Weight_without_magazine,Weight_with_empty_magazine,...,Dimension_LENGTH (OVERALL)*,Dimension_SLIDE LENGTH,Dimension_WIDTH (OVERALL),Dimension_SLIDE WIDTH,Dimension_HEIGHT INCL. MAG.,Dimension_,Dimension_Sight Radius POLYMER,Dimension_Sight Radius STEEL,Dimension_Sight Radius GNS,Dimension_TRIGGER DISTANCE*
0,G19,The G19 Gen5 is now available as an A-CUT™ COA...,https://eu.glock.com/en/products/pistols/g19-g...,9 mm Luger,Safe Action®,Standard: 15,102 mm | 4.02 inch,26 N,without magazine | 600 g | 21.16 oz incl. COA,with empty magazine | 680 g | 23.99 oz incl. COA,...,185 mm | 7.28 inch,174 mm | 6.85 inch,34 mm | 1.34 inch,"25,5 mm | 1.00 inch",128 mm | 5.04 inch,148 mm | 5.83 inch,NaN,146 mm | 5.75 inch,NaN,70 mm | 2.76 inch
1,G43X,The G43X is now available as an A-CUT™ COA™ Co...,https://eu.glock.com/en/products/pistols/g43x-...,9 mm Luger,Safe Action®,Standard: 10,87 mm | 3.43 inch,24 N,without magazine | 480 g | 16.93 oz incl. COA,with empty magazine | 545 g | 19.22 oz incl. COA,...,165 mm | 6.50 inch,154 mm | 6.06 inch,28 mm | 1.10 inch,22 mm | 0.87 inch,128 mm | 5.04 inch,148 mm | 5.83 inch,NaN,126 mm | 4.96 inch,NaN,67 mm | 2.64 inch
2,G45,The G45 is now available as an A-CUT™ COA™ Com...,https://eu.glock.com/en/products/pistols/g45-a...,9 mm Luger,Safe Action®,Standard: 17,102 mm | 4.02 inch,26 N,without magazine | 620 g | 21.87 oz incl. COA,with empty magazine | 700 g | 24.69 oz incl. COA,...,189 mm | 7.44 inch,174 mm | 6.85 inch,34 mm | 1.34 inch,"25,5 mm | 1.00 inch",139 mm | 5.47 inch,159 mm | 6.26 inch,NaN,146 mm | 5.75 inch,NaN,70 mm | 2.76 inch


A quick preview of the raw data in a tabular format, let's take a look at how many firearms there are and the list of column/attribute names to see what we're working with, lets also do a bit of clean up on those names.

In [3]:
print('Number of handpistols in dataset:',(len(data)))
print(data.columns)
print(len(data))

Number of handpistols in dataset: 86
Index(['name', 'description', 'url', 'Caliber', 'System', 'Mag. Capacity',
       'Barrel Length', 'Trigger Pull***', 'Weight_without_magazine',
       'Weight_with_empty_magazine', 'Weight_with_loaded_magazine',
       'Dimension_LENGTH (OVERALL)*', 'Dimension_SLIDE LENGTH',
       'Dimension_WIDTH (OVERALL)', 'Dimension_SLIDE WIDTH',
       'Dimension_HEIGHT INCL. MAG.', 'Dimension_',
       'Dimension_Sight Radius POLYMER', 'Dimension_Sight Radius STEEL',
       'Dimension_Sight Radius GNS', 'Dimension_TRIGGER DISTANCE*'],
      dtype='object')
86


In [4]:
# Column name clean up
data.rename(columns={
    'Mag. Capacity': 'mag_size',
    'Barrel Length': 'barrel_length',
    'Trigger Pull***': 'trigger_pull',
    'Dimension_LENGTH (OVERALL)*' : 'pistol_length', 
    'Dimension_SLIDE LENGTH' : 'slide_length',
    'Dimension_WIDTH (OVERALL)' : 'pistol_width', 
    'Dimension_SLIDE WIDTH' : 'slide_width',
    'Dimension_HEIGHT INCL. MAG.' : 'height_incl_mag', 
    'Dimension_' : 'height_incl_mag_2', # manually verified this attribute to be height including mag
    'Dimension_Sight Radius POLYMER' : 'sight_radius_polymer', 
    'Dimension_Sight Radius STEEL' : 'sight_radius_steel',
    'Dimension_Sight Radius GNS' : 'sight_radius_gns', 
    'Dimension_TRIGGER DISTANCE*' : 'trigger_distance',
}, inplace=True)

data.columns = map(str.lower, data.columns)

(data.head(1))


,name,description,url,caliber,system,mag_size,barrel_length,trigger_pull,weight_without_magazine,weight_with_empty_magazine,...,pistol_length,slide_length,pistol_width,slide_width,height_incl_mag,height_incl_mag_2,sight_radius_polymer,sight_radius_steel,sight_radius_gns,trigger_distance
0,G19,The G19 Gen5 is now available as an A-CUT™ COA...,https://eu.glock.com/en/products/pistols/g19-g...,9 mm Luger,Safe Action®,Standard: 15,102 mm | 4.02 inch,26 N,without magazine | 600 g | 21.16 oz incl. COA,with empty magazine | 680 g | 23.99 oz incl. COA,...,185 mm | 7.28 inch,174 mm | 6.85 inch,34 mm | 1.34 inch,"25,5 mm | 1.00 inch",128 mm | 5.04 inch,148 mm | 5.83 inch,NaN,146 mm | 5.75 inch,NaN,70 mm | 2.76 inch


### Each row is a pistol and every pistol has the following attributes

name - pistol name, pistols can have the same name but have different variants based on generation or the accessories they come with.

variant - __to be added later__, pistols can have variants being sold that include different features, accessories etc. (e.g. Glock-17, Glock-17C, Glock-17-T-MOS)

description - the pistols description as per the glock product page

url - the url that the data was collected from

caliber - the ammunition which the pistol chamber is designed to accommodate

system - the safety system of the pistol

mag_size - number of bullets the magazine or clip can hold at once

barrel_length - length of the pistol barrel, longer lengths generally improve accuracy over distance and higher muzzle velocities. (more power, less wind drift, )

trigger_pull - force required to pull the trigger

weight - three values for pistol weight with no magazine, empty magazine or loaded magazine.

height_incl_mag & height_incl_mag_2 - height including magazine, some pistols can be purchased with a different magazine hence the second value

sight_radius for steel/polymer/gns - sight radius indicates distance between front and rear sights, longer is better for accuracy over distance. glock offers sights in steel, polymer and night mode.

trigger_distance - distance needed for the trigger to fire the pistol.


In [5]:
data.head(1)

,name,description,url,caliber,system,mag_size,barrel_length,trigger_pull,weight_without_magazine,weight_with_empty_magazine,...,pistol_length,slide_length,pistol_width,slide_width,height_incl_mag,height_incl_mag_2,sight_radius_polymer,sight_radius_steel,sight_radius_gns,trigger_distance
0,G19,The G19 Gen5 is now available as an A-CUT™ COA...,https://eu.glock.com/en/products/pistols/g19-g...,9 mm Luger,Safe Action®,Standard: 15,102 mm | 4.02 inch,26 N,without magazine | 600 g | 21.16 oz incl. COA,with empty magazine | 680 g | 23.99 oz incl. COA,...,185 mm | 7.28 inch,174 mm | 6.85 inch,34 mm | 1.34 inch,"25,5 mm | 1.00 inch",128 mm | 5.04 inch,148 mm | 5.83 inch,NaN,146 mm | 5.75 inch,NaN,70 mm | 2.76 inch


In [6]:
exclude = ['description','url']
columns = [col for col in data.columns if  col not in exclude]

for col in columns:
    print(data[col].head(1))

0    G19
Name: name, dtype: object
0    9 mm Luger
Name: caliber, dtype: object
0    Safe Action®
Name: system, dtype: object
0    Standard: 15
Name: mag_size, dtype: object
0    102 mm | 4.02 inch
Name: barrel_length, dtype: object
0    26 N
Name: trigger_pull, dtype: object
0    without magazine | 600 g | 21.16 oz incl. COA
Name: weight_without_magazine, dtype: object
0    with empty magazine | 680 g | 23.99 oz incl. COA
Name: weight_with_empty_magazine, dtype: object
0    with loaded magazine** | 865 g | 30.51 oz incl...
Name: weight_with_loaded_magazine, dtype: object
0    185 mm | 7.28 inch
Name: pistol_length, dtype: object
0    174 mm | 6.85 inch
Name: slide_length, dtype: object
0    34 mm | 1.34 inch
Name: pistol_width, dtype: object
0    25,5 mm | 1.00 inch
Name: slide_width, dtype: object
0    128 mm | 5.04 inch
Name: height_incl_mag, dtype: object
0    148 mm | 5.83 inch
Name: height_incl_mag_2, dtype: object
0    NaN
Name: sight_radius_polymer, dtype: object
0    146 mm | 

In [7]:
columns_to_clean = {
    'barrel_length': 'barrel_length_mm',
    'trigger_pull': 'trigger_pull_N',
    'pistol_length': 'pistol_length_mm',
    'slide_length': 'slide_length_mm',
    'pistol_width': 'pistol_width_mm',
    'slide_width': 'slide_width_mm',
    'height_incl_mag': 'height_incl_mag_mm',
    'height_incl_mag_2': 'height_incl_mag_2_mm',
    'sight_radius_polymer': 'sight_radius_polymer_mm',
    'sight_radius_steel': 'sight_radius_steel_mm',
    'sight_radius_gns': 'sight_radius_gns_mm',
    'weight_without_magazine': 'weight_without_magazine_g',
    'weight_with_empty_magazine': 'weight_with_empty_magazine_g',
    'weight_with_loaded_magazine': 'weight_with_loaded_magazine_g',
    'trigger_distance': 'trigger_distance_mm'
}

# Process each column
for old_col, new_col in columns_to_clean.items():
    if old_col in data.columns:
        data[new_col] = data[old_col].str.extract(r'(\d+)').astype(float)
        data.drop(columns=[old_col], inplace=True)

# Process mag sizes
data['mag_size_standard'] = data['mag_size'].str.extract(r'Standard:\s*(\d+)').astype(float).astype(float)
data['mag_size_optional'] = data['mag_size'].str.extract(r'Optional:\s*(\d+)').astype(float).astype(float)
data.drop(columns=['mag_size'], inplace=True)

# Step 1: Create the 'variant' column
data['variant'] = data['url'].str.extract(r'pistols/[^-]+-(.*)')

# Step 2: Extract 'gen' from 'variant'
data['gen'] = data['variant'].str.extract(r'(gen\d)')

# Step 3: Remove 'gen' from 'variant' if present
# This replaces things like 'gen5-' at the beginning OR anywhere (with or without dash)
data['variant'] = data['variant'].str.replace(r'gen\d-?', '', regex=True)


cols = list(data.columns)
# Ensure 'variant' is right after 'name'
cols.insert(cols.index('name') + 1, cols.pop(cols.index('variant')))
# Move 'gen' right after 'variant'
cols.insert(cols.index('variant') + 1, cols.pop(cols.index('gen')))
data = data[cols]



data.head(5)


,name,variant,gen,description,url,caliber,system,barrel_length_mm,trigger_pull_N,pistol_length_mm,...,height_incl_mag_2_mm,sight_radius_polymer_mm,sight_radius_steel_mm,sight_radius_gns_mm,weight_without_magazine_g,weight_with_empty_magazine_g,weight_with_loaded_magazine_g,trigger_distance_mm,mag_size_standard,mag_size_optional
0,G19,a-cut-coa-combo,gen5,The G19 Gen5 is now available as an A-CUT™ COA...,https://eu.glock.com/en/products/pistols/g19-g...,9 mm Luger,Safe Action®,102.0,26.0,185.0,...,148.0,NaN,146.0,NaN,600.0,680.0,865.0,70.0,15.0,NaN
1,G43X,a-cut-coa-combo,NaN,The G43X is now available as an A-CUT™ COA™ Co...,https://eu.glock.com/en/products/pistols/g43x-...,9 mm Luger,Safe Action®,87.0,24.0,165.0,...,148.0,NaN,126.0,NaN,480.0,545.0,670.0,67.0,10.0,NaN
2,G45,a-cut-coa-combo,NaN,The G45 is now available as an A-CUT™ COA™ Com...,https://eu.glock.com/en/products/pistols/g45-a...,9 mm Luger,Safe Action®,102.0,26.0,189.0,...,159.0,NaN,146.0,NaN,620.0,700.0,910.0,70.0,17.0,NaN
3,G47,a-cut-coa-combo,NaN,The G47 is now available as an A-CUT™ COA™ Com...,https://eu.glock.com/en/products/pistols/g47-a...,9 mm Luger,Safe Action®,114.0,26.0,202.0,...,159.0,NaN,158.0,NaN,640.0,720.0,930.0,70.0,17.0,NaN
4,G48,a-cut-coa-combo,NaN,The G48 is now available as an A-CUT™ COA™ Com...,https://eu.glock.com/en/products/pistols/g48-a...,9 mm Luger,Safe Action®,106.0,24.0,185.0,...,148.0,NaN,146.0,NaN,540.0,605.0,730.0,67.0,10.0,NaN


In [23]:
categories = ['name', 'caliber', 'system', 'variant', 'gen']

for category in categories:
    print('')
    print(len(data[category].unique()),category,'values')
    print(data[category].unique())
    print('')



21 name values
['G19' 'G43' 'G45' 'G47' 'G48' 'G17' 'G18' 'G26' 'G34' 'G44' 'G22' 'G23'
 'G24' 'G27' 'G35' 'G20' 'G29' 'G21' 'G30' 'G25' 'G42']


7 caliber values
['9 mm Luger' '22 l.r.' '40 S&W' '10 mm Auto' '45 Auto' '380 Auto'
 'Training ammunition']


2 system values
['Safe Action®' 'Safe Action®\nSemi-/Full-automatic firing mode']


15 variant values
['a-cut-coa-combo' 'base' 'mos' 'l' 'c' 'gns' 'rail' 'mos-hunter-edition'
 'silver-slide' 'cut' 'p' 't-fx-fof' 't-mos-fx-fof' 't-mos-utm' 't-utm']


3 gen values
['gen5' nan 'gen4']



Categories seem good sensible, change the names of the pistols to reflect the base models, differences should be in the 'variant' column which I will add as well.

Additionally, there is an empty value '' in variant, which means some Glocks don't have a variant, we'll replace these empty values with "base" indicating they're the standard version of the Glock.

In [9]:
# Adding a column to set pistols with no variant to say base (base model)
data['variant'] = data['variant'].replace('', 'base').fillna('base')
data['variant'].unique()

# Modifying original names so they have the same base models (ie G17 and G17L both are glock 17s)
data['name'] = data['name'].str[:3]

In [10]:
# Looking into what glock models have Semi / Full auto mode
data[data['system']=='Safe Action®\nSemi-/Full-automatic firing mode']

,name,variant,gen,description,url,caliber,system,barrel_length_mm,trigger_pull_N,pistol_length_mm,...,height_incl_mag_2_mm,sight_radius_polymer_mm,sight_radius_steel_mm,sight_radius_gns_mm,weight_without_magazine_g,weight_with_empty_magazine_g,weight_with_loaded_magazine_g,trigger_distance_mm,mag_size_standard,mag_size_optional
12,G18,base,NaN,As a true fully automatic pistol the G18 is on...,https://eu.glock.com/en/products/pistols/g18,9 mm Luger,Safe Action®\nSemi-/Full-automatic firing mode,114.0,25.0,204.0,...,NaN,165.0,164.0,163.0,625.0,705.0,940.0,72.0,19.0,17.0
13,G18,c,NaN,The GLOCK G18C comes with compensator slots in...,https://eu.glock.com/en/products/pistols/g18-c,9 mm Luger,Safe Action®\nSemi-/Full-automatic firing mode,114.0,25.0,204.0,...,NaN,165.0,164.0,163.0,590.0,670.0,905.0,72.0,19.0,17.0


Everything else but the Glock 18 is semi-automatic.

After looking into the URLs and watching some <a href="https://www.youtube.com/shorts/ku9iIYZmki8">demo videos</a>, 

The Glock 18 is actually a machine pistol- fully automatic.

It seems that this can only be purchased by law enforcement and militaries- reasonable.

In [11]:
# Lastly, let's do some quick checks to see if there are anymore null values, duplicate values- or just attributes that need to be modified
data.head(1)

,name,variant,gen,description,url,caliber,system,barrel_length_mm,trigger_pull_N,pistol_length_mm,...,height_incl_mag_2_mm,sight_radius_polymer_mm,sight_radius_steel_mm,sight_radius_gns_mm,weight_without_magazine_g,weight_with_empty_magazine_g,weight_with_loaded_magazine_g,trigger_distance_mm,mag_size_standard,mag_size_optional
0,G19,a-cut-coa-combo,gen5,The G19 Gen5 is now available as an A-CUT™ COA...,https://eu.glock.com/en/products/pistols/g19-g...,9 mm Luger,Safe Action®,102.0,26.0,185.0,...,148.0,NaN,146.0,NaN,600.0,680.0,865.0,70.0,15.0,NaN


In [12]:
print(('Number of duplicates:'),data.duplicated().sum())
print("")
print(data.dtypes)


Number of duplicates: 0

name                              object
variant                           object
gen                               object
description                       object
url                               object
caliber                           object
system                            object
barrel_length_mm                 float64
trigger_pull_N                   float64
pistol_length_mm                 float64
slide_length_mm                  float64
pistol_width_mm                  float64
slide_width_mm                   float64
height_incl_mag_mm               float64
height_incl_mag_2_mm             float64
sight_radius_polymer_mm          float64
sight_radius_steel_mm            float64
sight_radius_gns_mm              float64
weight_without_magazine_g        float64
weight_with_empty_magazine_g     float64
weight_with_loaded_magazine_g    float64
trigger_distance_mm              float64
mag_size_standard                float64
mag_size_optional               

No duplicates, and after going through each attribute, the data types seem to be in order, cleaning and transforming done- we will move more into exploration

## Data Exploration and Analysis

In [26]:
sns.set_style('darkgrid')

plt.figure(figsize=(10, 6))
sns.set_style('darkgrid')
sns.countplot(x='caliber', data=data, 
              order=data['caliber'].value_counts().index)
plt.xticks(rotation=45)
plt.title("Bar Plot of Product Calibers")
plt.tight_layout()
plt.savefig('images/figures/caliber_count.png', bbox_inches='tight', dpi=300)
plt.close() 

<div align=center>
<h2> Analysis of Most Commonly Used Ammunition </h2>
<img src="./images/figures/caliber_count.png" alt="Distribution of Calibers" height="400"/>  

</div>

<h3>9mm Luger</h3>

Despite being portrayed as weak in pop culture, 9mm is highly effective in real-world use due to its low recoil, high magazine capacity, and strong stopping power with modern ammunition. These traits make it ideal for military, law enforcement, and self-defense, which explains why the majority of Glock pistols in the dataset are chambered in 9mm. Glock’s original Glock 17 helped popularize the caliber, and with NATO standardization and advances in bullet design, most agencies now favor 9mm over heavier rounds.


<h3>.40 S&W</h3>

Originally designed for law enforcement in response to concerns about 9mm’s stopping power, .40 S&W delivers more energy and penetration, but with heavier recoil and lower capacity. It became widely used in the '90s but has since declined as 9mm ammo improved over time. The chart reflects this shift—Glock produces far fewer .40-caliber models today, aligning with a broader industry move back to 9mm for its balance for performance and efficiency.

<h3>Training Ammunition</h3>

Certain glocks are training models and are designed to be used with special training ammunition. Typically including plastic, rubber, paint, low-velocity rounds.

<p align="center">
<img src="./images/several-ammo.jpg" alt="Comparison of 9mm and 40 S&W and Training Ammo" height="300"/>
</p>







In [27]:
print("Heaviest Glock")
print(data.sort_values(by='weight_without_magazine_g', ascending=False)[['name','variant', 'weight_with_loaded_magazine_g', 'caliber']].head(1))

print("")

print("Lightest Glock")
print(data.sort_values(by='weight_without_magazine_g', ascending=False)[['name','variant', 'weight_with_loaded_magazine_g', 'caliber']].tail(1))
print("\nProduct Description: ",data.iloc[73]['description'])


Heaviest Glock
   name variant  weight_with_loaded_magazine_g caliber
56  G24    base                         1090.0  40 S&W

Lightest Glock
   name variant  weight_with_loaded_magazine_g   caliber
73  G42    base                          450.0  380 Auto

Product Description:  For those discerning shooters who desired the smallest GLOCK possible, the company responded with the now popular G42 in 380 AUTO with ultra-compact dimensions and easy-to-shoot characteristics. The G42 packs all the features GLOCK customers have come to appreciate - from the GLOCK Safe Action System with its multiple integrated safety features to the brand’s proven reliability and durability.


<div align="center">
  <b><h2 align=center>The Heaviest and Lightest of Glocks</h2></b>
  <img src="./images/G24vsG42.jpg" alt="Comparison of Biggest and Smallest Glock" height="300"/>  
</div>

  <h3>The Heaviest Glock - Glock 24</h3>
  Product Description: <i>The long slide and barrel design combined with the 40 S&W caliber cartridge make this GLOCK a great choice for a multitude of shooting activities.</i>
  <br><br>
  In other words, this is much more performant than your average glock, chambered in a .40 S&W instead of the typical 9mm as mentioned earlier, it features a longer muzzle and an overall bigger frame, making it easier to aim over distance, introducing less bullet wobble and a more aggressive flight path when compared to more "typical" models. After doing some research, it turns out that this is a fan-favorite glock, <a href="https://www.thetruthaboutguns.com/glocks-forgotten-six-inchers">primarily designed for competition shooting.</a> It is also regarded as one of the rarest glocks.
  
  <h3>The Lightest Glock - Glock 42</h3>
  Product Description: <i>For shooters who desired the smallest GLOCK possible, the company responded with the now popular G42 in 380 AUTO with ultra-compact dimensions and easy-to-shoot characteristics.</i>
  <br><br>
  Essentially the opposite of the previous model, so small that the whole pistol can essentially fit in the palm of your hand, chambered in 380 ACP which is a very small round compared to the 9mm and 40 S&W. It's a practical everyday carry for self-defense, individuals with smaller hands might also prefer this kind of pistol, not very powerful but power is not the selling point of this product.
 
  <h3>.40 S&W VS .380 ACP</h3>
  The .40 S&W is a much larger and more powerful cartridge than the .380 ACP. While the .40 S&W offers greater stopping power and longer effective range, the .380 ACP provides less recoil, making it easier to control, especially in compact firearms. This difference in size and power directly influences the design and intended use of both the Glock 24 and Glock 42 models, where the Glock 24 is commonly used by competition shooters, and the Glock 42 used for concealed carry and self defense, by individuals with smaller handsizes such as women.


In [28]:
order = data['name'].value_counts().sort_values(ascending=False).index

sns.set_style('darkgrid')

plt.figure(figsize=(10, 6))
sns.set_style('darkgrid')
sns.countplot(x='name', data=data, order=order)
plt.xticks(rotation=45)
plt.title("Count Plot of Models by Name")
plt.tight_layout()
plt.savefig('images/figures/model_distribution', bbox_inches='tight', dpi=300)
plt.close() 


<div align=center><img src="./images/figures/model_distribution.png" alt="Count Plot of Product Models" height="500"/> </div>

Over a third of glocks available for purchase are variants of the Glock 17 and the Glock 19, upon asking around and doing my own research, the similarities between the two models are so strong that the 19 is essentially just a slightly smaller version of the 17. Both are chambered in 9mm, have nearly identical trigger pulls, controls and accuracy potential.

<div align=center>
<img src="./images/G17vsG19.jpg" alt="Glock 17 and 19 Comparison" height="400"/>
</div>

The major differences lie in the dimensions and capacity, the Glock 19 is slightly smaller and lighter with a mag capacity of 15 instead of 17, making it a bit better for concealed carry- but people still use the Glock 17 for this as well.


In [29]:
# Return Glock 17 and 19 base models for analysis
g17andg19 = data[
    ((data['name'] == 'G17') | (data['name'] == 'G19')) & (data['variant'] == 'base')
][[
    'name',
    'variant',
    'gen',
    'caliber',
    'system',
    'weight_without_magazine_g',
    'weight_with_loaded_magazine_g',
    'mag_size_standard'
]].sort_values('weight_with_loaded_magazine_g')
g17andg19

,name,variant,gen,caliber,system,weight_without_magazine_g,weight_with_loaded_magazine_g,mag_size_standard
14,G19,base,NaN,9 mm Luger,Safe Action®,600.0,855.0,15.0
15,G19,base,gen4,9 mm Luger,Safe Action®,600.0,855.0,15.0
17,G19,base,gen5,9 mm Luger,Safe Action®,610.0,855.0,15.0
44,G19,base,gen5,9 mm Luger,Safe Action®,610.0,855.0,15.0
5,G17,base,NaN,9 mm Luger,Safe Action®,625.0,915.0,17.0
6,G17,base,gen4,9 mm Luger,Safe Action®,625.0,915.0,17.0
8,G17,base,gen5,9 mm Luger,Safe Action®,630.0,945.0,17.0
41,G17,base,gen5,9 mm Luger,Safe Action®,630.0,945.0,17.0


As mentioned, the base G19 and G17 models are both chambered in 9mm, and have quite low weights being loaded or unloaded. 

What makes each model special in Glock's lineup then?

The Glock 17 wasn't just Glock’s first pistol—it was also the company’s first firearm of any kind. Gaston Glock had previously manufactured curtain rods and knives for the military. Despite industry skepticism over its polymer frame, the G17 passed NATO durability tests without a single malfunction. Lesser known is that early Glock 17s lacked a manual safety, sparking controversy and contributing to its “plastic gun” myth. Over time, the G17 evolved into the most diversified Glock, with training, threaded-barrel, and MOS (Modular Optic System) variants—making it both an innovator and a platform.

The Glock 19, introduced in 1988, was Glock’s second pistol and the company’s first compact model. Building upon the success of the full-size Glock 17, the G19 offered a shorter barrel and grip, making it more suitable for concealed carry while maintaining a respectable 15-round magazine capacity. This balance of size and firepower quickly made it a favorite among law enforcement, military personnel, and civilian shooters alike. Its versatility allowed it to serve effectively as both a duty weapon and a concealed carry firearm just like its predecessor.

In [30]:
plt.figure(figsize=(10, 6))
sns.histplot(data['weight_with_loaded_magazine_g'], bins=13, kde=True)
plt.xticks(rotation=45)
plt.title("Distribution of Product Weights (Loaded)")
plt.tight_layout()
plt.savefig('images/figures/distribution_gun_weight', bbox_inches='tight', dpi=300)
plt.close() 

<div align=center><img src="./images/figures/distribution_gun_weight.png" alt="Count Plot of Product Models" height="500"/> </div>

We can see that across the Glock product range, a majority of their models sit around the 900g mark when paired with a stock-issued loaded magazine, on the low end we see the G42, weighing in at just under half a kilogram, and on the high end we see models like the G24 which weigh just over a kilogram. This is generally considered on the low to medium end for pistol weights, visually highlights Glock's philosophy of capitalizing on polymer-frames to reduce product weight, making them more ideal to carry around when compared to competitors products which are traditionally heavier and more troublesome to carry concealed.

<h2 align=center>Glock 17 (Gen 5) vs Beretta 92FS</h2>

<div align=center>
<img src="./images/G17vsM9.jpg" alt="Glock 17 vs  M9" height="400"/>
</div>

<br>

<table align=center border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse;">
    <tr>
        <th style="border: 1px solid black;">Feature</th>
        <th style="border: 1px solid black;">Glock 17 Gen 5</th>
        <th style="border: 1px solid black;">Beretta 92FS</th>
    </tr>
    <tr>
        <td style="border: 1px solid black;">Unloaded Weight</td>
        <td style="border: 1px solid black;">630 grams</td>
        <td style="border: 1px solid black;">945 grams</td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">Chambering</td>
        <td style="border: 1px solid black;">9mm</td>
        <td style="border: 1px solid black;">9mm</td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">Magazine Capacity</td>
        <td style="border: 1px solid black;">17 rounds</td>
        <td style="border: 1px solid black;">15 rounds</td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">Frame Material</td>
        <td style="border: 1px solid black;">Polymer-based</td>
        <td style="border: 1px solid black;">Aluminum</td>
    </tr>
</table>


The Glock 17 and Beretta 92FS are both full-size 9mm service pistols that serve as flagship models in their respective lineups. Despite both being chambered in 9mm with similar magazine capacities, they differ notably in design and market positioning.

The Glock 17 like many other glocks, features a polymer frame, contributing to its lighter weight of approximately 630 grams when unloaded. In contrast, the Beretta 92FS, with its aluminum alloy frame, weighs about 945 grams unloaded . Nearly a 50% difference in product weight, which highlights the benefits of Glock products for extended carry, appealing to both law enforcement and civilian users seeking a balance between portability and performance.

Both pistols hold significant positions in their manufacturers' offerings. The Glock 17, originally introduced in the 1980s, set the standard for Glock's line of reliable, easy-to-maintain firearms. The Beretta 92FS, adopted by the U.S. military as the M9 in 1985, has been a symbol of durability and precision in Beretta's lineup.

In terms of market share, Glock has achieved substantial coverage, particularly in U.S. law enforcement. A research survey sent across the US law enforcement agencies in 2018 showcases that <a href="https://www.officer.com/tactical/firearms/press-release/21001909/glock-inc-a-growing-number-of-us-law-enforcement-agencies-make-the-switch-to-glock-pistols?utm_source=chatgpt.com"> Glock pistols were the preferred choice for 65% of U.S. law enforcement agencies. </a> This widespread adoption underscores Glock's emphasis on lightweight design, reliability, and cost-effectiveness, distinguishing it from competitors like Beretta.

In summary, while both the Glock 17 and Beretta 92FS are esteemed service pistols, the Glock 17, like many of the other product offerings, utilizes its lighter weight and smaller size to appeal strongly to its customerbase, especially among law enforcement agencies seeking efficiency and ease of use.

## Final Thoughts: Glock’s Unique Market Position

<div align=center>
<img src="./images/closing-glock-art.jpg" alt="G19 Gen 4" height="400"/>
</div>

Glock's story is one of disruptive innovation. A company that entered the firearms market without traditional gunsmithing experience and redefined the rules through engineering efficiency and polymer technology. Its reputation for simplicity, reliability, and durability has made it a dominant force in law enforcement, military, and civilian sectors alike.

Despite being occasionally underestimated in popular media as a token firearm, Glock pistols consistently deliver professional-grade performance across a wide range of use cases, for a wide range of shooters. The data analysis I've conducted reveals a product line that is not only consistent in quality and design philosophy, but also diverse in function and design but also, reinforcing Glock’s status as a leader in the firearms industry.

Glock’s unique market positioning stems from its ability to balance functionality with manufacturing precision, all while maintaining an iconic brand identity. It serves as a compelling example of how engineering-first thinking and product reliability can carve out lasting dominance in a competitive and highly scrutinized industry.


In [1]:
#python -m nbconvert --to webpdf --embed-images EDA.ipynb